virtio-devices/src/block.rs

```diff
@@ -116,6 +121,34 @@ impl BlockEpollHandler {
 
         for avail_desc in queue.iter(&mem) {
             let mut request = Request::parse(&avail_desc, &mem).map_err(Error::RequestParsing)?;
+
+            // If limiter.consume() fails it means there is no more TokenType::Ops
+            // budget and rate limiting is in effect.
+            if !self.rate_limiter.consume(1, TokenType::Ops) {
+                // Stop processing the queue and return this descriptor chain to the
+                // avail ring, for later processing.
+                queue.go_to_previous_position();
+                break;
+            }
+            // Exercise the rate limiter only if this request is of data transfer type.
+            if request.request_type == RequestType::In || request.request_type == RequestType::Out {
+                let mut bytes = Wrapping(0);
+                for (_, data_len) in &request.data_descriptors {
+                    bytes += Wrapping(*data_len as u64);
+                }
+
+                // If limiter.consume() fails it means there is no more TokenType::Bytes
+                // budget and rate limiting is in effect.
+                if !self.rate_limiter.consume(bytes.0, TokenType::Bytes) {
+                    // Revert the OPS consume().
+                    self.rate_limiter.manual_replenish(1, TokenType::Ops);
+                    // Stop processing the queue and return this descriptor chain to the
+                    // avail ring, for later processing.
+                    queue.go_to_previous_position();
```

Comment:
Based on the fact that I'd prefer `self.rate_limiter` to be `Option<RateLimiter>`, this entire block would only be executed in case:
```rust
if let Some(rate_limiter) = &self.rate_limiter {
    ...
```

Reply:
The block of code looks good to me, with `go_to_previous_position()` being invoked in the right places.